### PR TITLE
Migrate reentrancy

### PR DESCRIPTION
**Pull request by jtriley-eth**
## Changes

The following changes have been made:

- Reentrancy library has been migrated from the std library to sway-libs
- Reentrancy tests have been migrated
- A new directory, `test_artifacts` has been introduced for multi-contract tests
- The build script (`.github/scripts/build.sh`) has been modified to build `test_artifacts`

## Notes

- It may be worth reviewing the rest of the ci pipeline to ensure the new directory is handled properly

## Related Issues

Part 1 of [Issue 3217 in Fuellabs/sway](https://github.com/FuelLabs/sway/issues/3217)

### CODE REVIEW

**Review by nfurfaro — COMMENTED** [review]
The changes look good to me. Another set of eyes on the CI-related things is warranted.

**Review by matt-user — APPROVED** [review]
I looked through the successful ci run and everything seems to be in order.  The test project is built along with the test artifacts and then the tests are run.

File: tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw
Comment (Braqzen) [inline tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw:52]:
```suggestion
        let attacker = abi(Attacker, contract_id().value);
```

what about this?

File: tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw
Comment (Braqzen) [inline tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw:26]:
```suggestion
        let target = abi(Target, target.value);
```

No reason to have an additional variable here

File: tests/src/test_projects/reentrancy/mod.rs
Comment (Braqzen) [inline tests/src/test_projects/reentrancy/mod.rs:13]: I would at least separate these into successful tests and those that revert modules plus another one to contain the setup. This makes it easier to navigate because chunks of code are contained together

File: tests/src/test_projects/reentrancy/mod.rs
Comment (Braqzen) [inline tests/src/test_projects/reentrancy/mod.rs:31]: `Revert(0)` comes from the `assert()` failing. Are these tests correct?

File: tests/src/test_projects/reentrancy/mod.rs
Comment (Braqzen) [inline tests/src/test_projects/reentrancy/mod.rs:81]: I would factor out these strings into constant variables defined at the top and used here

File: tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw
Comment (jtriley-eth) [reply tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw]: Abstracted this away by just using `get_msg_sender_id_or_panic()` and removed the explicit type annotation.

File: sway_libs/src/reentrancy/reentrancy.sw
Comment (jtriley-eth) [reply sway_libs/src/reentrancy/reentrancy.sw]: Removed

File: tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw
Comment (jtriley-eth) [reply tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw]: I don't think so, this was the original test implementation. I've updated this and the respective ABIs.

File: tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw
Comment (jtriley-eth) [reply tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw]: Removed single-use variables

File: tests/src/test_projects/reentrancy/mod.rs
Comment (jtriley-eth) [reply tests/src/test_projects/reentrancy/mod.rs:19]: What would be a good structure for this, something like the following?

- `tests/src/test_projects/reentrancy/mod.rs`
- `tests/src/test_projects/reentrancy/pass.rs`
- `tests/src/test_projects/reentrancy/fail.rs`

File: tests/src/test_projects/reentrancy/mod.rs
Comment (jtriley-eth) [reply tests/src/test_projects/reentrancy/mod.rs]: Done.

File: tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw
Comment (nfurfaro) [reply tests/src/test_artifacts/reentrancy_attacker_contract/src/main.sw]: I had to return a bool originally due to limitations on what I could pass into and return from abi functions from the SDK.

**Review by nfurfaro — COMMENTED** [review]
Looks good. Will approve once all comments are resolved and checks are green.

**Pull request closed without merging**
